### ISSUE TITLE
`TypeAliasType` is mistakenly considered assignable to the type it represents

### ISSUE THREAD

**Issue by InSyncWithFoo** [microsoft/pyright#8444]
Minimal reproducible example (playgrounds: [Pyright](https://pyright-play.net/?pyrightVersion=1.1.371&pythonVersion=3.13&strict=true&code=C4TwDgpgBAgg2gFQLpQLxQQWAFA4CYQBmUhAFAG4BcscAzsAE5ICUUAtAHxQByA9gHYRKOAJAA6CThxl49JqygBiKP15QIDBrwZA), [Mypy](https://mypy-play.net/?flags=strict&mypy=master&python=3.12&enable-incomplete-feature=NewGenericSyntax&gist=58eebc4116e4dc534568104c0194badf)):

```python
type A[T] = T

def f(v: A[str]) -> None:
	...

f(A[str])  # pyright => no error
           # mypy => error: Argument 1 to "f" has incompatible type "TypeAliasType"; expected "str"
```

**Comment by erictraut** [comment]
I think pyright is doing the right thing here. The type of `A` when used in a value expression is `TypeAliasType`. The typeshed definition of this class defines a `__getitem__` method which returns `Any`. That means `A[0]` evaluates to `Any`, and that explains why you're not seeing an error in this case.

It would be theoretically possible to override the behavior indicated in the typeshed stub, but I generally avoid this. The few times that I've done this have resulted in downstream problems including unexpected side effects and maintainability issues.

For example, the following code is arguably legitimate and should not generate a type violation.

```pyright
type A[T] = T
x: type[str] = A[str][0]
```

**Comment by erictraut** [comment]
@JelleZijlstra, do you have any thoughts on this?

**Comment by JelleZijlstra** [comment]
At runtime, `TypeAliasType.__getitem__` always returns a GenericAlias, so I submitted a typeshed PR (linked above) to reflect that behavior in the stub. With that change, the OP's code should yield a pyright error.
